File: backend/src/controllers/order_controller.js
const { default: mongoose } = require("mongoose");
const orderModel = require("../models/order_model");

const orderController = {
    createOrder: async function (req, res) {
        try {
            const { items, seller, buyer, } = req.body;
            console.log(items);
            console.log(" details :  " + buyer);
            const foundOrder = await orderModel.findOne({ buyer: buyer });
            // console.log(foundCart);
            if (foundOrder == null) {
                console.log(foundOrder);
                const creatingOrder = new orderModel({ seller: seller, buyer: buyer, items: items });
                // creatingOrder.items.push({  productid:product ,quantity:quantity
                // });
                await creatingOrder.save();
                return res.json({ success: true, message: "Order Successfully placed", data: creatingOrder });
            }
            //  await orderModel.findOneAndUpdate(//first pull then push the item in db
            //     {buyer:buyer,"items.product":product}, //match user and product 
            //     {$pull:{items:{product:product,}} }, //dollar pull is array function ->pull value from array means delete values from array using function pull
            //     );
            //if cart exists
            // items.forEach(element => {
            // console.log(element.orderedProduct.productid);
            // console.log(element.totalquantity);
            // });
            items[0].orderedProduct.forEach(element => {
                console.log(element);
            });
            const updatedOrder = await orderModel.findOneAndUpdate(
                { buyer: buyer }, //match condition
                { $push: { items: items } }, // make a changes                //dollar push ->push value in  array function pull
                { new: true } // new version of doc
            ).populate("items.orderedProduct.productid");
            // const updatedOrder =  await orderModel.findOneAndUpdate(
            //     {buyer:buyer}, //match condition
            //     {$push:{items:{productid:product,quantity:quantity}} }, // make a changes                //dollar push ->push value in  array function pull
            //     {new: true} // new version of doc
            //     ).populate("items.product");

            return res.json({ success: true, message: "Order Successfully placed", data: updatedOrder });

        } catch (ex) {
            return res.json({ success: false, message: `Order can't Placed :  ${ex}` });

        }
    },

    fetchOrders: async function (req, res) {

        try {
            const buyerid = req.params.buyerId;
            console.log("orders ");
            console.log(buyerid);

            const foundOrders = await orderModel.find(
                { buyer: buyerid },
            ).populate("items.orderedProduct.productid");
            // .populate('items.orderedProduct.buyer');
            // .populate("items.product").populate("seller");
            console.log(foundOrders);
            console.log("we");

            if (!foundOrders) {
                // console.log(foundBid);
                return res.json({ success: false, data: [], message: "There is no orders to display" });
            } else
                return res.json({ success: true, message: "order found successfully", data: foundOrders });
        } catch (ex) {
            return res.json({ success: false, message: ex.message });
        }
    },

    fetchordersForSeller: async function (req, res) {
        try {
            const sellerId = req.params.sellerId;

            const foundbids = await orderModel.find({
                // "items.orderedProduct.seller": sellerId
            },
                // {totalprice:0,totalquantity:0}
            ).populate('items.orderedProduct.productid').populate('items.orderedProduct.buyer');

            console.log(foundbids);

            if (!foundbids) {


                return res.json({ success: false, data: [], message: "There is no item in your bid" });
            } else {


                const filteredBids = foundbids.map(order => {
                    const filteredItems = order.items.map(item => {
                        const filteredOrderedProducts = item.orderedProduct.filter(product => {
                            return product.seller._id == sellerId;
                        });
                        return {
                            ...item._doc,
                            orderedProduct: filteredOrderedProducts
                        };
                    }).filter(item => item.orderedProduct.length > 0);
                    return {
                        ...order._doc,
                        items: filteredItems
                    };
                }).filter(order => order.items.length > 0);


                return res.json({ success: true, message: "bid found", data: filteredBids });

            }
        } catch (ex) {
            return res.json({ success: false, message: `"bid not found ${ex}"` });

        }
    },

    updateOrderStatus: async function (req, res) {
        try {
            const { status, orderId, buyerId } = req.body;
            console.log("buyerid " + buyerId + " orderId: " + orderId + " status: " + status);

            const order = await orderModel.findOneAndUpdate(
                {
                    buyer: buyerId,
                    'items._id': orderId
                },
                {
                    $set: {
                        'items.$.status': status
                    }
                },
                { new: true }
            );

            console.log(order);
            if (order == null) {
                return res.json({ success: false, message: "order not found." });

            }
            return res.json({ success: true, message: "order found and updated status.", data: status });


        } catch (ex) {
            return res.json({ success: false, message: `"bid not found ${ex}"` });

        }
    }


};





module.exports = orderController;

File: Fyp/bidbazar/lib/controllers/order_controller.dart
import 'package:bidbazar/controllers/auth_controllers.dart';
import 'package:bidbazar/controllers/cart_controller.dart';
import 'package:bidbazar/controllers/product_controller.dart';
import 'package:bidbazar/data/models/order_model.dart';
import 'package:bidbazar/data/models/product_model.dart';
import 'package:bidbazar/data/models/user_model.dart';
import 'package:bidbazar/data/repo/order_repo.dart';
import 'package:bidbazar/data/repo/product_repo.dart';
import 'package:dio/dio.dart';
import 'package:get/get.dart';

class OrderController extends GetxController with StateMixin {
  Rx<OrderModel> order = OrderModel().obs;
  OrderRepo orderRepo = OrderRepo();
  List<OrderedProduct> orderProducts = [];
  productRepo productrepo = productRepo();
  RxList<Items> orderList = (List<Items>.of([])).obs;
  product_controller productController = Get.put(product_controller());
  RxString orderStatus="Order Placed".obs;
  RxList<Items> OrdersFilters=<Items>[].obs;
  
  @override
  void onClose() {
    // TODO: implement onClose
    super.onClose();
  }

  @override
  void onInit() async {
    // TODO: implement onInit
    super.onInit();
  }
  Future pendingOrderFilter()async{
    for (var element in orderList) {
      if(element.status=="Pending"){
       OrdersFilters.add(element);
      }
    }
  }
   Future deliveredOrderFilter()async{

     for (var element in orderList) {
      if(element.status=="Delivered"){
        OrdersFilters.add(element);
      }
    }
  }
   Future canceldOrderFilter()async{
    for (var element in orderList) {
      if(element.status=="Canceld"){
       OrdersFilters.add(element);
      }
    }
  }

  Future placeOrder(
      {required double totalprice, required int totalquantity}) async {
    try {
      change(order.value, status: RxStatus.loading());

      cartController controller = Get.put(cartController());
      for (var element in controller.cartlist) {
        print("Start working");
        orderProducts.add(OrderedProduct(
            productid: productModel(
                sId: element.product!.sId!, user: element.product!.user!),
            quantity: element.quantity!,
            buyer: AuthenticateController.userdata.first,
            seller: userModel(sId: element.product!.user!),
            bidprice: element.product!.price!
            ),
            );
        //product sold added
        productrepo.updateSoldQty(
            productId: element.product!.sId!, soldqty: element.quantity!);
        print("product qty updated ");
      }
      print("order product works ");
      print(orderProducts.length);

      // convert order item into json format
      // List<Map<String, dynamic>> orderedProductsJson = orderProducts.map((product) => product.toJson()).toList();
      print("complete order product works ");

      await orderRepo.createOrder(
        items: [
          Items(
            orderedProduct: orderProducts,
            totalquantity: totalquantity,
            totalprice: totalprice.toInt(),
          )
        ],
        buyerId: AuthenticateController.userdata.first.sId!,
        //  sellerId: sellerId
      );
      // if(!hasdata){
      // change(order, status: RxStatus.error("There is no item to bid"));
      //   }
      await productController.fetchProducts();
      productController.productList.refresh();
      change(order.value, status: RxStatus.success());
    } on DioException catch (ex) {
      change(order.value, status: RxStatus.error(ex.toString()));
    }
  }

  Future fetchOrders() async {
    orderList.clear();
    OrdersFilters.clear();
    try {
      change(orderList, status: RxStatus.loading());

      List<OrderModel> orders = await orderRepo.fetchOrders(
          Id: AuthenticateController.userdata.first.sId!);

      print("working order data");
      print(orders);
      // orderList.assignAll(orders);

      if (orders.isEmpty) {
        change(orderList, status: RxStatus.empty());
      } else {
        // for (var i in orders){
        orderList.addAll(orders[0].items!);

        change(orderList, status: RxStatus.success());
      }
    } catch (ex) {
      change(orderList, status: RxStatus.error(ex.toString()));
    }
  }

 Future orderStatusUpdate({required String status,required String orderId,required String buyerId}) async {
    try {

     var order =   orderRepo.orderStatus(status: status, orderId: orderId, buyerId: buyerId);
         print(order);
    } catch (ex) {
      change(orderList, status: RxStatus.error(ex.toString()));
    }
  }
  //   Future updateBidStatus({required String buyerId,required String productId,required String status}) async {
  //         await bid.updateBidStatus(buyerId:buyerId , productId: productId, status: status) ;
  // }
}
